### ISSUE TITLE
(Experimental Workflow Interface) Aggregator private attributes visible to Collaborators 

### ISSUE THREAD

**Issue by refai06** [intel/openfl#1052]
**Describe the bug** 
In the Experimental aggregator-based workflow, private attributes of the aggregator object are passed to the collaborator step. This behavior does not occur in `LocalRuntime`.

**To Reproduce** 
Steps to reproduce the behavior in the experimental setup:

1.  Create and set up the workspace with 301_torch_cnn_mnist_watermarking
     **Path**: `openfl/openfl-workspace/experimental/301_torch_cnn_mnist_watermarking`
2.  Modify the flow step: `aggregated_model_validation` as follows: 

```python
    @collaborator
    def aggregated_model_validation(self):
        """
        Perform Aggregated Model validation on Collaborators.
        """
        print(‘***************** Debug ****************’)
        print vars(self)
        print(‘***************** Debug ****************’)
        self.agg_validation_score = inference(self.model, self.test_loader)
        print(f"<Collab: {self.input}>"
              + f" Aggregated Model validation score = {self.agg_validation_score}"
              )

        self.next(self.train)
```

3.  Run Aggregator and Collaborator
4.  Observe that self of the collaborator contains the aggregator’s private attributes (as shown below): 

Logs
```shell
{

.
.
.
.
.
aggregator_private_attribute:
'pretrain_epochs': 25,
'retrain_epochs': 25,
'watermark_acc_threshold': 0.98,
'watermark_data_loader': <torch.utils.data.dataloader.DataLoader object at 0x7ff088619610>,
'watermark_pretraining_completed': True
.
.
.
.

}
```
**Expected behavior**
The aggregator’s private attributes should not be visible in the collaborator step.

**Comment by MasterSkepticista** [comment]
Is this solvable using the `exclude` keyword argument in a flow step?

**Comment by scngupta-dsp** [comment]
> Is this solvable using the `exclude` keyword argument in a flow step?

@MasterSkepticista: Participant private attributes are not meant to be filtered by using exclude keyword arguments inside the flow steps. Workflow Interface should ensure that aggregator private attributes are always filtered out of the current state when transferring from aggregator to collaborators

**Comment by refai06** [comment]
Issue was analysed in detail and following problems were found in Aggregator based workflow (FederatedRuntime):

 

#1: For FederatedRuntime, clones are created in the next method of the FLSpec class

After the aggregator executes the start step, FLSpec object contains the aggregator’s private attributes. Since clones are created from this object, aggregator private attributes are also passed to the clone objects

 

#2: For FederatedRuntime, the filter_exclude_include functionality is called in the next method of the FLSpec class during the transition from aggregator to collaborator

This results in same issue as above. As the FLSpec object already contains the aggregator private attributes therefore these are again filtered into the clones

**Comment by ishant162** [comment]
Closed with #1084
